### PR TITLE
[548] Expose Redis operations via Redis trait

### LINKED ISSUE #548: Introduce all Redis operations as type-safe methods of RedisExecutor trait
In order to facilitate better use by layers, which want to consume other services by their interfaces, it would be ideal to add to `RedisExecutor` typed methods for all Redis operations, which simply encode/decode as appropriate.

This way, a service can depend on RedisExecutor, and use its operations, without also directly adding any dependencies into ZIO environment (which would have to be implemented for a service which is calling RedisExecutor).

### PR DESCRIPTION
**Pull request by m-kalai**
Resolves #548 

* getting rid of `Redis` environment dependency from all the API traits
* mixing API traits in the `Redis` trait
* getting rid of "accessors" from `redis` package object and fixing tests

### CODE REVIEW

**Comment by mijicd** [comment]
Am I missing something? This PR seems to contain only benchmarks.

**Comment by m-kalai** [comment]
@mijicd you scared me there a bit :) but i see all the changes, not just in the benchmark module

**Comment by mijicd** [comment]
> @mijicd you scared me there a bit :) but i see all the changes, not just in the benchmark module

I guess it's GitHub app issue. I'll check it out from laptop.

File: redis/src/main/scala/zio/redis/CommandExecutor.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package zio.redis
+
+import zio.schema.codec.BinaryCodec
+
+trait CommandExecutor {
+  implicit def codec: BinaryCodec
+
+  implicit def executor: RedisExecutor
+}
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/CommandExecutor.scala:9]: Let's rename it somehow, as this is not really an executor.

File: redis/src/main/scala/zio/redis/api/Lists.scala [hunk 2]
```diff
@@ -20,10 +20,10 @@ import zio._
 import zio.redis.Input._
 import zio.redis.Output._
 import zio.redis.ResultBuilder._
-import zio.redis._
+import zio.redis.{CommandExecutor, _}
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/api/Lists.scala:23]: This change can be reverted.

File: redis/src/main/scala/zio/redis/codecs/CRC16.scala [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package zio.redis.codec
+package zio.redis.codecs
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/codecs/CRC16.scala:1]: This change is not really related to the PR scope, but I like it so let it stay.

File: redis/src/test/scala/zio/redis/ClusterExecutorSpec.scala [hunk 4]
```diff
@@ -9,53 +9,63 @@ object ClusterExecutorSpec extends BaseSpec {
...
-          _                = ZIO.logDebug(s"$key _____ Migrating $keySlot from ${sourceMaster.id}- ${sourceMaster.address}")
-          sourceMasterConn = getRedisNodeLayer(sourceMaster.address)
-          _               <- setSlotMigrating(keySlot, destMaster.id).provideLayer(sourceMasterConn)
-          value2          <- get(key).returning[String] // have to redirect without error ASK
-          value3          <- get(key).returning[String] // have to redirect without creating new connection
-          _               <- setSlotStable(keySlot).provideLayer(destMasterConn)
-        } yield {
-          assertTrue(value1 == value2) && assertTrue(value2 == value3)
+        ZIO.serviceWithZIO[Redis] { redis =>
+          import redis._
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/ClusterExecutorSpec.scala:13]: Let's not do this.

File: redis/src/test/scala/zio/redis/ClusterExecutorSpec.scala [hunk 5]
```diff
@@ -9,53 +9,63 @@ object ClusterExecutorSpec extends BaseSpec {
...
-          _ <- migrate(destMaster.address.host, destMaster.address.port.toLong, key, 0, 5.seconds, keys = None)
-                 .provideLayer(sourceMasterConn)
-          _      <- setSlotNode(keySlot, destMaster.id).provideLayer(destMasterConn)
-          _      <- setSlotNode(keySlot, destMaster.id).provideLayer(sourceMasterConn)
-          value2 <- get(key).returning[String] // have to refresh connection
-          value3 <- get(key).returning[String] // have to get value without refreshing connection
-        } yield {
-          assertTrue(value1 == value2) && assertTrue(value2 == value3)
+        ZIO.serviceWithZIO[Redis] { redis =>
+          import redis._
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/ClusterExecutorSpec.scala:39]: Same here.

File: redis/src/test/scala/zio/redis/ConnectionSpec.scala [hunk 6]
```diff
@@ -12,85 +12,97 @@ trait ConnectionSpec extends BaseSpec {
...
       suite("clientCaching")(
         test("track keys") {
-          for {
-            _            <- clientTrackingOff
-            _            <- clientTrackingOn(trackingMode = Some(ClientTrackingMode.OptIn))
-            _            <- clientCaching(true)
-            trackingInfo <- clientTrackingInfo
-          } yield assert(trackingInfo.flags.caching)(isSome(isTrue))
+          ZIO.serviceWithZIO[Redis] { redis =>
+            import redis._
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/ConnectionSpec.scala:16]: Same here.

File: redis/src/test/scala/zio/redis/ConnectionSpec.scala [hunk 7]
```diff
@@ -12,85 +12,97 @@ trait ConnectionSpec extends BaseSpec {
...
         },
         test("don't track keys") {
-          for {
-            _            <- clientTrackingOff
-            _            <- clientTrackingOn(trackingMode = Some(ClientTrackingMode.OptOut))
-            _            <- clientCaching(false)
-            trackingInfo <- clientTrackingInfo
-          } yield assert(trackingInfo.flags.caching)(isSome(isFalse))
+          ZIO.serviceWithZIO[Redis] { redis =>
+            import redis._
```
Comment (mijicd) [inline redis/src/test/scala/zio/redis/ConnectionSpec.scala:27]: Same here.

File: example/src/main/scala/example/contributorsCache.scala [hunk 8]
```diff
@@ -36,12 +36,15 @@ final case class ContributorsCacheLive(r: Redis, s: Sttp) extends ContributorsCa
...
       .ignore
 
   private def read(repository: Repository): ZIO[Redis, ApiError, Contributors] =
-    get(repository.key)
-      .returning[String]
+    ZIO
+      .serviceWithZIO[Redis](
+        _.get(repository.key)
+          .returning[String]
+      )
```
Comment (mijicd) [inline example/src/main/scala/example/contributorsCache.scala:47]:
```suggestion
    ZIO.serviceWithZIO[Redis](_.get(repository.key).returning[String])
```

**Review by mijicd — CHANGES_REQUESTED** [review]
Almost there :)

File: redis/src/main/scala/zio/redis/RedisCommand.scala [hunk 9]
```diff
@@ -30,6 +30,12 @@ final class RedisCommand[-In, +Out] private (val name: String, val input: Input[
       }
       .refineToOrDie[RedisError]
 
+  private[redis] def run(in: In)(implicit executor: RedisExecutor, codec: BinaryCodec): ZIO[Any, RedisError, Out] =
+    executor
+      .execute(resp(in, codec))
+      .flatMap[Any, Throwable, Out](out => ZIO.attempt(output.unsafeDecode(out)(codec)))
+      .refineToOrDie[RedisError]
+
```
[the remark this answers is not in the document]
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/RedisCommand.scala:38]: I don't insist on implicits, it was just a way to get things done faster. I can make them "explicit" args or constructor params. Makes probably more sense as a constructor param as you suggested. Any particular order? Second argument list?

[the remark this answers is not in the document]
Comment (mijicd) [reply redis/src/main/scala/zio/redis/RedisCommand.scala:38]: Sure. It probably makes sense grouping it that, but I'm also fine if you place them in the same one.

File: redis/src/main/scala/zio/redis/codecs/CRC16.scala (hunk 3 above)
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/codecs/CRC16.scala:1]: yeah you're right, sorry about that. i renamed it because at certain point i had a collision of codec package and codec method on redis package object, and i forgot about it. i'll keep it then.

File: redis/src/main/scala/zio/redis/CommandExecutor.scala (hunk 1 above)
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/CommandExecutor.scala:9]: yeah i was struggling with naming this one, considering those two methods were making Redis trait before this change. could it be something like `RedisEnvironment` or `CommandExecutable`. Any suggestion?

Comment (mijicd) [reply redis/src/main/scala/zio/redis/CommandExecutor.scala:9]: Environment sounds good. They don't need to be implicit anymore either. The question remains: can we remove it completely?

Comment (m-kalai) [reply redis/src/main/scala/zio/redis/CommandExecutor.scala]: i don't think we can, unless we want to split command creation and execution. If we changed API traits to return `RedisCommand` and pushed invocation of `run` then we wouldn't need these dependencies on those traits. Or do you have some ideas?

File: redis/src/main/scala/zio/redis/RedisEnvironment.scala [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package zio.redis
+
+import zio.schema.codec.BinaryCodec
+
+trait RedisEnvironment {
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RedisEnvironment.scala:5]: Maybe we should make it package-private. After all, it is an implementation detail.

File: redis/src/main/scala/zio/redis/RedisCommand.scala [hunk 11]
```diff
@@ -20,27 +20,25 @@ import zio._
...
+  private[redis] def resp(in: In): Chunk[RespValue.BulkString] =
     Varargs(StringInput).encode(name.split(" "))(codec) ++ input.encode(in)(codec)
 }
 
 object RedisCommand {
-  private[redis] def apply[In, Out](name: String, input: Input[In], output: Output[Out]): RedisCommand[In, Out] =
-    new RedisCommand(name, input, output)
+  private[redis] def apply[In, Out](name: String, input: Input[In], output: Output[Out])(
+    executor: RedisExecutor,
+    codec: BinaryCodec
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/RedisCommand.scala:41]: I'm not sure about this parameter-group. Let's merge them in the first one (sorry :pray: )

**Review by mijicd — COMMENTED** [review]
Looks really good to me. Two minor things and it's ready to go.

File: redis/src/main/scala/zio/redis/CommandExecutor.scala (hunk 1 above)
Comment (mijicd) [reply redis/src/main/scala/zio/redis/CommandExecutor.scala]: I would not "expose" run. As mentioned below, 2nd parameter group is a bit too much as well, imo. On the other side, it makes sense if we restore implicits. :thinking:

File: redis/src/main/scala/zio/redis/RedisEnvironment.scala (hunk 10 above)
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/RedisEnvironment.scala]: yes, makes sense

File: redis/src/main/scala/zio/redis/RedisCommand.scala (hunk 11 above)
Comment (m-kalai) [reply redis/src/main/scala/zio/redis/RedisCommand.scala]: done

File: redis/src/main/scala/zio/redis/CommandExecutor.scala (hunk 1 above)
Comment (mijicd) [reply redis/src/main/scala/zio/redis/CommandExecutor.scala]: That's not bad. We might be able to combine it with `returning` somehow and make a nice "proxy", but let's leave it for the follow-up PR (if you're willing to dig into it :) ).

**Comment by mijicd** [comment]
@m-kalai Please adjust the docs examples and it's ready to go.

**Review by mijicd — APPROVED** [review]

**Review by mijicd — APPROVED** [review]
